src/posix/platform/resolver.cpp

```diff
@@ -0,0 +1,293 @@
+                mUpstreamDnsServerList[mUpstreamDnsServerCount] = addr;
+                mUpstreamDnsServerCount++;
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void Resolver::Query(otPlatDnsUpstreamQuery *aTxn, const otMessage *aQuery)
+{
+    char        packet[kMaxDnsMessageSize];
+    otError     error  = OT_ERROR_NONE;
+    uint16_t    length = otMessageGetLength(aQuery);
+    sockaddr_in serverAddr;
+
+    Transaction *txn;
+
+    VerifyOrExit(length <= kMaxDnsMessageSize, error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(otMessageRead(aQuery, 0, &packet, sizeof(packet)) == length, error = OT_ERROR_NO_BUFS);
+
+    txn = AllocateTransaction(aTxn);
+    VerifyOrExit(txn != nullptr, error = OT_ERROR_NO_BUFS);
+
+    serverAddr.sin_family = AF_INET;
+    serverAddr.sin_port   = htons(53);
+    for (int i = 0; i < mUpstreamDnsServerCount; i++)
+    {
+        serverAddr.sin_addr.s_addr = mUpstreamDnsServerList[i];
+        VerifyOrExit(sendto(txn->mUdpFd, packet, length, 0, (struct sockaddr *)&serverAddr, sizeof(serverAddr)),
+                     error = OT_ERROR_NO_ROUTE);
+    }
+    otLogInfoPlat("Forwarded DNS query %p to %d server(s).", static_cast<void *>(aTxn), mUpstreamDnsServerCount);
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogCritPlat("Failed to forward DNS query %p to server: %d", static_cast<void *>(aTxn), error);
+    }
+    return;
+}
+
+void Resolver::Cancel(otPlatDnsUpstreamQuery *aTxn)
+{
+    Transaction *txn = GetTransaction(aTxn);
+
+    CloseTransaction(txn);
+}
+
+Resolver::Transaction *Resolver::AllocateTransaction(otPlatDnsUpstreamQuery *aThreadTxn)
+{
+    int          i;
+    int          error     = 0;
+    int          fdOrError = 0;
+    Transaction *ret       = nullptr;
+
+    for (i = 0; i < kMaxUpstreamTransactionCount; i++)
+    {
+        if (mUpstreamTransaction[i].mThreadTxn == nullptr)
+        {
+            ret       = &mUpstreamTransaction[i];
+            fdOrError = socket(AF_INET, SOCK_DGRAM, 0);
+            if (fdOrError < 0)
+            {
+                otLogInfoPlat("Failed to create socket: %d", fdOrError);
+                error = fdOrError;
+                break;
+            }
+            ret->mUdpFd                        = fdOrError;
+            mUpstreamTransaction[i].mThreadTxn = aThreadTxn;
+            break;
+        }
+    }
+
+    if (ret != nullptr && error != 0)
+    {
+        CloseTransaction(ret);
+        ret = nullptr;
+    }
+
+    return ret;
+}
+
+void Resolver::ForwardResponse(Transaction *aTxn)
+{
+    char       response[kMaxDnsMessageSize];
+    ssize_t    readSize;
+    otMessage *message;
+
+    VerifyOrExit((readSize = read(aTxn->mUdpFd, response, sizeof(response))) > 0);
+
+    message = otUdpNewMessage(gInstance, nullptr);
+    SuccessOrExit(otMessageAppend(message, response, readSize));
+
+    otPlatDnsUpstreamQueryDone(gInstance, aTxn->mThreadTxn, message);
+    message = nullptr;
+
```

Comment:
Should we  free/close the `aTxn` (and the related `fd`?)
Call `CloseTransaction()`?

Reply:
Added CloseTransaction in Process